**Exploring non-African archaic segments**

In this project, you will be looking at segments of archaic genomes identified in individual modern humans. You will investigate how alike non-African genomes are in terms of how archaic segments are distributed. You will be working with an extended version of the data set that you worked with in the admixture exercise. In this version you also have an addtional file with the positions of candidate archaic SNPs.

**Investigate the following**

In your project, you must address the questions below, but you are also expected to expand the project to answer your own questions. How you do this is up to you. You do not need to answer them in the order they are listed. Make a project plan with a set of analyses that will allow you to answer the questions.


```{r}
#loading in the necessary packages
library(correlation)
library(tidyverse)
library(corrr)
library(ggplot2)
library(reshape2)
library(RColorBrewer)
library(ggnewscale)
library(colorspace)
```

I start out using the SNP dataset so I can ceck which snps each individual has.
I filter out the "human" which is the snps that were not found in one of the Neanderthal or Denisova genomes.
```{r}
# Loading in the SNP dataset
snps <- read.table("SNPs.txt",header=TRUE, sep="\t", na.strings="NA", dec=".", strip.white=TRUE)

# Keep only archaic SNPs (exclude "human")
snp_archaic <- snps %>%
  filter(archaic != "human") %>%
  distinct(name, chrom, start, .keep_all = TRUE) %>%  # Keep only one row per SNP per individual
  arrange(region)

# A vector of all the snp's in the dataset since it could come in handy later
unique_snps <- unique(snp_archaic$start)

# Create a list of unique individuals. This will be used later for plotting the data
individuals <- unique(snp_archaic$name)

snp_denisova <- snps %>%
  filter(archaic == "Denisova_nofil") %>%
  distinct(name, chrom, start, .keep_all = TRUE) %>%  # Keep only one row per SNP per individual
  arrange(region)

snp_neanderthal <- snps %>%
  filter(archaic == "AltaiNea_nofil" | archaic =="Vi33.19_nofil") %>%
  distinct(name, chrom, start, .keep_all = TRUE) %>%  # Keep only one row per SNP per individual
  arrange(region)
```

Here I prepare for checking the correlation in the snps that each individual contains. 
It takes some time to do for all of the individuals, but the results of it will be used throughout the rest of the analysis.
```{r}
# Make table with each individual as a column and each snp as a row. 
# Each cell is a 1 if the individual has the snp and a 0 if it doesn't
snp_matrix <- snp_archaic %>%
  select(name, start) %>%
# Step 2: Create presence/absence table
  distinct(name, start) %>%  # Remove duplicates if any
  mutate(present = 1) %>%
  pivot_wider(names_from = name, values_from = present, values_fill = 0)

snp_matrix_denisova <- snp_denisova %>%
  select(name, start) %>%
# Step 2: Create presence/absence table
  distinct(name, start) %>%  # Remove duplicates if any
  mutate(present = 1) %>%
  pivot_wider(names_from = name, values_from = present, values_fill = 0)


```


```{r}
# Makes a correlation matrix of a subset or of the whole dataset of snps and individuals.
cor_matrix <- snp_matrix %>%
  select(-start)%>%
  correlate()

cor_matrix_denisova <- snp_matrix_denisova %>%
  select(-start)%>%
  correlate()

```


```{r}
# loading in the segments dataset which contains information about geographic placement of individuals
segments <- read.table("ArchaicSegments.txt",header=TRUE, sep="\t", na.strings="NA", dec=".", strip.white=TRUE)

```

Here I get an overview of the ArchaicSegments data. Checking how many individuals there are from each population, country and region. It is rather useful since we don't want to look at correlation in populations with only one individual for example.
```{r}
# Count unique individuals per population
count_pop <- segments %>%
  distinct(name, pop) %>%
  count(pop, name = "individuals_per_pop")

# Count unique individuals per country
count_country <- segments %>%
  distinct(name, country) %>%
  count(country, name = "individuals_per_country")

# Count unique individuals per region
count_region <- segments %>%
  distinct(name, region) %>%
  count(region, name = "individuals_per_region")

```


**A. To what extent do individuals share SNPs contributed by archaic human introgression? In other words, how correlated are the archaic contents in two individuals?**

```{r}

# Function to merge ranges and calculate total length
merge_segments <- function(df) {
  
  ir <- IRanges(start = df$start, end = df$end)
  merged <- reduce(ir)  # Merge overlapping/adjacent segments
  total_length <- sum(width(merged))  # Sum the lengths
  
 
  
  return(total_length)
  
}
```

```{r}
library(IRanges)
# Summaries of each group
segments_ind <- segments %>%
  filter(length > 0) %>%
  group_by(name) %>%
  summarise(total_length = merge_segments(cur_data()),
            pop = unique(pop),
            country = unique(country),
            region = unique(region),
            name = unique(name),
            .groups = "drop")
  
 detach("package:IRanges", unload = TRUE, character.only = TRUE)
detach("package:S4Vectors", unload = TRUE, character.only = TRUE)
detach("package:BiocGenerics", unload = TRUE, character.only = TRUE)
```

This is for dividing the correlation data later
```{r}
# Makes a dataset which only contains each individual once and their population, country and region
segments_clean <- segments %>%
  distinct(name, pop, country, region)
```

```{r}
# Saving the correlation matrix as a txt-file for later use. Important for the analysis with all of the individuals
write.table(cor_matrix, file = "shared_snps_matrix.txt", sep = "\t", quote = FALSE)
```

```{r}
# Reads in the full correlation matrix of all the individuals
cor_matrix <- read.table("shared_snps_matrix_all.txt", check.names = FALSE)
```

```{r}
# Make the matrix into a tibble with each row as a correlation between two individuals
cor_tibble <- melt(cor_matrix, value.name = "correlation") %>%
  rename(ind1 = term, ind2 = variable) # Rename columns if needed

# Apply the factor levels with uppercase names
cor_tibble <- cor_tibble %>%
  mutate(
    ind1 = factor(ind1, levels = individuals),
    ind2 = factor(ind2, levels = individuals)
  )

# Make the matrix into a tibble with each row as a correlation between two individuals
cor_tibble_denisova <- melt(cor_matrix_denisova, value.name = "correlation") %>%
  rename(ind1 = term, ind2 = variable) # Rename columns if needed

# Apply the factor levels with uppercase names
cor_tibble_denisova <- cor_tibble_denisova %>%
  mutate(
    ind1 = factor(ind1, levels = individuals),
    ind2 = factor(ind2, levels = individuals)
  )
```

```{r}
# Join the correlation data with the metadata about the origin of the individuals
cor_annotated <- cor_tibble %>%
  left_join(segments_clean, by = c("ind1" = "name")) %>%
  left_join(segments_clean, by = c("ind2" = "name"), suffix = c("_1", "_2"))

cor_annotated_denisova <- cor_tibble_denisova %>%
  left_join(segments_clean, by = c("ind1" = "name")) %>%
  left_join(segments_clean, by = c("ind2" = "name"), suffix = c("_1", "_2"))
```


```{r}
# Create a named vector mapping individual names to their index
name_to_index <- setNames(individuals_df$index, individuals_df$individual)

# Add index to cor_annotated (you might already have this, but ensure it's sorted)
cor_annotated <- cor_annotated %>%
  mutate(
    ind1_index = name_to_index[ind1],
    ind2_index = name_to_index[ind2]
    )

name_to_index <- setNames(individuals_df$index, individuals_df$individual)

# Add index to cor_annotated (you might already have this, but ensure it's sorted)
cor_annotated_denisova <- cor_annotated_denisova %>%
  mutate(
    ind1_index = name_to_index[ind1],
    ind2_index = name_to_index[ind2]
    )
```


```{r}
ggplot(cor_annotated, aes(x = ind1_index, y = ind2_index, fill = correlation)) +
  geom_tile() +

  # First fill scale: correlations
  scale_fill_gradient2(high = "white", low = "darkblue", mid = "#EE2C2C", midpoint = 0.5,
                       name = "Correlation") +
  new_scale_fill() +

  # Top annotation
  geom_tile(
    data = individuals_df,
    aes(x = index, y = 0, fill = region),
    inherit.aes = FALSE,
    height = 10
  ) +

  # Left annotation
  geom_tile(
    data = individuals_df,
    aes(x = 0, y = index, fill = region),
    inherit.aes = FALSE,
    width = 10
  ) +

  # Region color scale
  scale_fill_manual(values = region_palette, name = "Region") +

  # Theme
  theme_minimal() +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    axis.title = element_blank(),
    panel.grid = element_blank(),
    plot.margin = margin(20, 60, 20, 60),
    legend.position = "right"
  ) +
  coord_fixed(clip = "off")

```


This makes a dataset of the mean correlations between individuals in each population. Note that some populations in the dataset only contains one individual and is therefore not taken into account.
```{r}
# Correlations between individuals from the same populations
same_pop <- cor_annotated %>%
  filter(pop_1 == pop_2) %>%
  group_by(pop_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1), # Count unique individuals
    country = unique(country_1) 
  ) %>%
  filter(num_individuals > 1)  # Skip populations with only 1 individual

mean_pop <- mean(same_pop$mean_correlation)

mean_all <- mean(na.omit(cor_annotated_denisova$correlation))
```


**B. How does this correlation change when you compare individuals from different populations?**

Since we only want to know how the correlation changes when we compare individuals from different populations I decided to make it so we only individuals that are from the same country, but from different populations.
```{r}
# Correlation between individuals from different populations, but still the same country
same_country <- cor_annotated %>%
  filter(country_1 == country_2)%>%
  group_by(country_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1)  # Count unique individuals
  ) %>%
  filter(num_individuals > 1)  # Skip countries with only 1 individual

mean_country <- mean(same_country$mean_correlation)
```


**C. How does it change when comparing individuals from different geographical regions?**

```{r}
same_region <- cor_annotated %>%
  filter(region_1 == region_2) %>%
  group_by(region_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1)) # Count unique individuals
    
mean_region <- mean(same_region$mean_correlation)
    
diff_region <- cor_annotated %>%
  filter(region_1 != region_2) %>%
  group_by(region_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1), # Count unique individuals
    
  ) %>%
  filter(num_individuals > 1)  # Skip regions with only 1 individual

mean_diff_region <- mean(diff_region$mean_correlation)

cor_annotated %>%
  filter(region_1 != region_2)
```

```{r}
bar_x <- c("All", "Same population", "Same country", "Different region")
bar_y <- c(mean_all, mean_pop, mean_country, mean_diff_region)

bar_data <- data.frame(x = bar_x, y = bar_y)

ggplot(bar_data, aes(x = x, y = y)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  labs(
    x = "Geographic partition",
    y = "Mean correlation in archaic SNPs",
    title = "Mean correlation in different subsets"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5)
  )

```


```{r}
create_heatmap <- function(region, data) {
  # Filter to the region of interest
  cor_region <- data %>%
    filter(region_1 == region, region_1 == region_2) %>%
    arrange(country_1, pop_1)  # Ensures individuals are ordered by country

  # Extract unique individuals and assign an index based on country-sorted order
  individuals_df <- data.frame(
    individual = cor_region$ind1,
    country = cor_region$country_1
  ) %>%
    unique() %>%
    arrange(country, individual) %>%
    mutate(index = row_number())

  # Create lookup vector: name -> index
  name_to_index <- setNames(individuals_df$index, individuals_df$individual)

  # Apply index to correlation data
  cor_region <- cor_region %>%
    mutate(
      ind1_index = name_to_index[ind1],
      ind2_index = name_to_index[ind2]
    )

  # Generate color palette for countries
  n <- length(unique(individuals_df$country))
  country_palette <- qualitative_hcl(n, palette = "Dark 3")
  names(country_palette) <- sort(unique(individuals_df$country))

  # Create the plot
  ggplot(cor_region, aes(x = ind1_index, y = ind2_index, fill = correlation)) +
    geom_tile() +
    scale_fill_gradient2(
      high = "white", low = "darkblue", mid = "#EE2C2C", midpoint = 0.5,
      name = "Correlation"
    ) +
    new_scale_fill() +

    # Top annotation
    geom_tile(
      data = individuals_df,
      aes(x = index, y = 0, fill = country),
      inherit.aes = FALSE,
      height = 0.5 + 0.5 * log(n)
    ) +

    # Left annotation
    geom_tile(
      data = individuals_df,
      aes(x = 0, y = index, fill = country),
      inherit.aes = FALSE,
      width = 0.5 + 0.5 * log(n)
    ) +

    scale_fill_manual(values = country_palette, name = "Country") +

    labs(title = region) +
    theme_minimal() +
    theme(
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      axis.title = element_blank(),
      panel.grid = element_blank(),
      plot.margin = margin(20, 60, 20, 60),
      legend.text = element_text(size = 6),
      legend.title = element_text(size = 7),
      legend.key.size = unit(0.4, "cm"),
      legend.position = "right",
      plot.title = element_text(size = 14, face = "bold", hjust = 0.5)
    ) +
    coord_fixed(clip = "off")
}

```


```{r}

regions <- unique(cor_annotated$region_1)

for (region in regions){
  heat <- create_heatmap(region, cor_annotated)
  print(heat)
}
```


**D. Does the region containing the EPAS1 gene stand out in any way? (Redo the analysis above for a 1Mb window surrounding this gene).**

```{r}
# Finding how much to add and subtract from the area to get 1Mb window
(1000000 - (46386697 - 46293667)) / 2

# Making the 1 Mb window around the EPAS1 gene
epas_window <- c(46293667 - 453485, 46386697 + 453485)
```

```{r}
# Makes a correlation matrix of a subset or of the whole dataset 
epas_snps_matrix <- snp_matrix %>%
  filter(start >= epas_window[1], start <= epas_window[2])

epas_cor_matrix <- epas_snps_matrix %>%
  correlate()
```

```{r}
# Make the matrix into a tibble with each row as a correlation between two individuals
epas_cor_tibble <- melt(epas_cor_matrix, value.name = "correlation") %>%
# Rename columns if needed
  rename(ind1 = term, ind2 = variable)
# Apply the factor levels with uppercase names
epas_cor_tibble <- epas_cor_tibble %>%
  mutate(
    ind1 = factor(ind1, levels = individuals),
    ind2 = factor(ind2, levels = individuals)
  )


# Join metadata
epas_cor_annotated <- epas_cor_tibble %>%
  left_join(segments_clean, by = c("ind1" = "name")) %>%
  left_join(segments_clean, by = c("ind2" = "name"), suffix = c("_1", "_2"))

for (reg in regions){
  epas_heat <- create_heatmap(region = reg, epas_cor_annotated)
  print(epas_heat)
}
```
```{r}
# Correlations between individuals from the same populations
epas_same_pop <- epas_cor_annotated %>%
  filter(pop_1 == pop_2) %>%
  group_by(pop_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1)  # Count unique individuals
  ) %>%
  na.omit() %>%
  filter(num_individuals > 1)  # Skip populations with only 1 individual

```

```{r}
# Correlation between individuals from different populations, but still the same country
epas_same_country <- epas_cor_annotated %>%
  filter(country_1 == country_2)%>%
  group_by(country_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1)  # Count unique individuals
  ) %>%
  na.omit() %>%
  filter(num_individuals > 1)  # Skip countries with only 1 individual


```

```{r}
epas_dif_region <- epas_cor_annotated %>%
  filter(region_1 != region_2) %>%
  group_by(region_1) %>%
  summarise(
    mean_correlation = mean(correlation, na.rm = TRUE),
    num_individuals = n_distinct(ind1)  # Count unique individuals
  ) %>%
  na.omit() %>%
  filter(num_individuals > 1)  # Skip regions with only 1 individual


```

```{r}
epas_mean_pop <- mean(epas_same_pop$mean_correlation)
epas_mean_all <- mean(na.omit(epas_cor_annotated$correlation))
epas_mean_country <- mean(epas_same_country$mean_correlation)
epas_mean_dif_region <- mean(epas_dif_region$mean_correlation)



epas_bar_x <- c("All", "Same population", "Same country", "Different region")
epas_bar_y <- c(epas_mean_all, epas_mean_pop, epas_mean_country, epas_mean_dif_region)

epas_bar_data <- data.frame(x = epas_bar_x, y = epas_bar_y)

ggplot(epas_bar_data, aes(x = x, y = y)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  labs(
    x = "Geographic partition",
    y = "Mean correlation in archaic SNPs",
    title = "Mean correlation in different subsets"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5)
  )
```

```{r}
combined_bar_data <- data.frame(
  Category = rep(bar_x, 2),
  Correlation = c(epas_bar_y, bar_y),
  Gene = rep(c("EPAS1", "All"), each = length(bar_x))
)

ggplot(combined_bar_data, aes(x = Category, y = Correlation, fill = Gene)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(
    x = "Geographic partition",
    y = "Mean correlation in archaic SNPs",
    title = "Comparison of mean correlation across subsets"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5)
  ) +
  scale_fill_manual(values = c("EPAS1" = "steelblue", "All" = "darkorange"))

```


**E. What is the total amount of admixture (archaic genomic sequence) in each non-African individual genome?**

Ranges from 1.5% to 4.4% with a mean of 2.7%
```{r}
segments_ind$percentage <- segments_ind$total_length/(3.1*10^9)
mean(segments_ind$percentage)
max(segments_ind$percentage)
min(segments_ind$percentage)

```


**F. What is the total amount of admixture (archaic genomic sequence) in the region around EPAS1 in each individual?**

Ranges from 0.8% to 91% with a mean of 40%
```{r}
library(IRanges)
epas_segments <- segments %>%
  filter(length > 0,start >= epas_window[1], end <= epas_window[2]) %>%
  group_by(name) %>%
  summarise(total_length = merge_segments(cur_data()),
            pop = unique(pop),
            country = unique(country),
            region = unique(region),
            name = unique(name),
            .groups = "drop")

epas_segments$percentage <- epas_segments$total_length / (epas_window[2] - epas_window[1])

detach("package:IRanges", unload = TRUE, character.only = TRUE)
detach("package:S4Vectors", unload = TRUE, character.only = TRUE)
detach("package:BiocGenerics", unload = TRUE, character.only = TRUE)

mean(epas_segments$percentage)
max(epas_segments$percentage)
min(epas_segments$percentage)
```


**G. Do individuals with large admixture totals have more correlated admixture patterns? Do individuals with large admixture totals in the EPAS1 region have more correlated admixture patterns in the EPAS1 region? Can you find any evidence of adaptive introgression?**

```{r}
top_10_percent <- segments_ind %>%
  filter(total_length >= quantile(total_length, 0.90, na.rm = TRUE))

low_10_percent <- segments_ind %>%
  filter(total_length <= quantile(total_length, 0.10, na.rm = TRUE))

top_10_corr <- cor_annotated %>%
  filter(ind1 %in% top_10_percent$name & ind2 %in% top_10_percent$name)

low_10_corr <- cor_annotated %>%
  filter(ind1 %in% low_10_percent$name & ind2 %in% low_10_percent$name)

mean(na.omit(top_10_corr$correlation))
mean(na.omit(low_10_corr$correlation))


```

```{r}
top_10_epas <- epas_segments %>%
  filter(total_length >= quantile(total_length, 0.90, na.rm = TRUE))

low_10_epas <- epas_segments %>%
  filter(total_length <= quantile(total_length, 0.10, na.rm = TRUE))

top_10_epas_corr <- epas_cor_annotated %>%
  filter(ind1 %in% top_10_epas$name & ind2 %in% top_10_epas$name)

low_10_epas_corr <- epas_cor_annotated %>%
  filter(ind1 %in% low_10_epas$name & ind2 %in% low_10_epas$name)

mean(na.omit(top_10_epas_corr$correlation))
mean(na.omit(low_10_epas_corr$correlation))
```


**H. Perform any additional analyses of your own choice.**



**Papers**

Analysis of Human Sequence Data Reveals Two Pulses of Archaic Denisovan Admixture
Altitude adaptation in Tibetans caused by introgression of Denisovan-like DNA


**Data**

Data for the project can be found in this folder on the cluster:
~/populationgenomics/project_data/ArchaicAdmixture
The files are:

ArchaicSegments.txt: This file is formatted the same way as the one you used for the archaic admixture exercise.
SNP.txt: This file has all the non-African SNPs that remain after removing all SNPs found in (Subsaharan) Africa. When each SNP is found in any of the high-coverage archaic genomes (Altai or Vindija Neanderthals or the Denisova), it is labeled as such. Otherwise, it is labeled "human".
